src/server/app/lib/App.js

```diff
@@ -0,0 +1,278 @@
+/*
+ * Copyright (c) 2012-2015 S-Core Co., Ltd.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * App class representing an App
+ * Use App.getInstance() to get an instance of App class.
+ *
+ * @since: 15. 11. 4
+ * @author: Koong Kyungmi (kyungmi.k@samsung.com)
+ */
+
+var url = require('url');
+var _ = require('underscore');
+var path = require('path');
+
+var utils = require('../../common/utils');
+var logger = require('../../common/log-manager');
+var config = require('../../common/conf-manager').conf;
+var db = require('../../common/db-manager')('app');
+var dao = db.dao;
+
+var ClientError = utils.ClientError;
+var ServerError = utils.ServerError;
+
+
+var rootApp = null;
+
+function App(appid) {
+    this.id = appid;
+    this.appid = appid;
+    this.domain = null;
+    this.apptype = null;
+    this.name = null;
+    this.desc = null;
+    this.ownerId = null;
+    this.status = 'stopped';
+}
+
+function getDomainByRequest(req) {
+    var host = req.host;
+    var domain = null;
+    logger.debug('getDomainByRequest request : ', req.host, req.originalUrl);
+    logger.info('deploy type', config.services.app.deploy.type);
+
+    // It has only exception for empty subdomain ('') because of default webida-client app.
+    // and deploy type 'path' always has the path started with 'pathPrefix'
+    // for making a distinction with the default app.
+    if (host === config.domain)  {
+        domain = '';
+    } else {
+        domain = host.substr(0, host.indexOf('.' + config.domain));
+    }
+    if (domain === '' && config.services.app.deploy.type === 'path') {
+        var paths = req.path.split('/');
+        if (paths.length > 2 && paths[1] === config.services.app.deploy.pathPrefix) {
+            domain = paths[2];
+        }
+    }
+
+    logger.debug('getDomainByRequest domain : ', domain);
+    return domain;
+}
+
+/**
+ * Set and return rootApp global variable
+ */
+function getRootApp(callback) {
+    if (rootApp) {
+        return callback(null, rootApp);
+    }
+    var app = new App('');
+    app.getAppInfo(function (err, appInfo) {
+        if (err) {
+            return callback(err);
+        }
+        if (!appInfo) {
```

Comment:
Do we have 'root app' in DB? Why do we need this? I think we can simplify App.js more, but not an urgent task.